src/Two/AbstractProvider.php

```diff
@@ -200,9 +200,9 @@ protected function hasInvalidState()
             return false;
         }
 
-        $session = $this->request->getSession();
+        $state = $this->request->getSession()->pull('state');
 
-        return ! ($this->request->input('state') === $session->get('state'));
+        return ! (strlen($state) > 0 && $this->request->input('state') === $state);
```

Comment:
Sorry to swoop in, I found this PR via [FriendsOfPhp](https://github.com/FriendsOfPHP/security-advisories/pull/78/files) and was just glancing around and wonder if the above comparison is susceptible to a [timing attack](https://en.wikipedia.org/wiki/Timing_attack)? If an attacker can control `$this->request->input('state')` then strict string equality may be leaking information.